File: go/leetcode/easy/645-错误的集合.go
package main

import (
	"math"
	"sort"
)

func findErrorNums(nums []int) []int {

	newSums := []int{}
	sort.Ints(nums)
	tempMap := make(map[int]bool)
	for _, i := range nums {
		tempMap[i] = true
	}
	trueSum := (len(nums) + 1) * len(nums) / 2
	setSum := 0
	sum := 0
	for m := range tempMap {
		newSums = append(newSums, m)
	}
	for _, ns := range newSums {
		setSum += ns
	}
	for _, num := range nums {
		sum += num
	}
	return []int{sum - setSum, trueSum - setSum}
}

// 方法很巧妙
func findErrorNums1(nums []int) []int {
	n := len(nums)
	res := make([]int, 2)
	for i := 0; i < n; i++ {
		index := int(math.Abs(float64(nums[i]))) - 1
		if nums[index] < 0 {
			res[0] = int(math.Abs(float64(nums[i])))
		} else {
			nums[index] *= -1
		}
	}
	for i := 0; i < n; i++ {
		if nums[i] > 0 {
			res[1] = i + 1
		}
	}
	return res
}

func main() {

}

File: python/leetcode/algorithms/easy/645-错误的集合.py
from typing import List


class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        newNums = sorted(nums)
        temp = 0
        miss = 0
        redundant = 0
        for i in newNums:
            if i - temp == 0:
                redundant = i
            elif i - temp > 1:
                miss = i - 1
            temp = i
        if miss == 0:
            if newNums[redundant - 1] == redundant:
                return [redundant, newNums[len(newNums) - 1] + 1]
            return [redundant, redundant + 1]
        return [redundant, miss]

    def findErrorNums1(self, nums: List[int]) -> List[int]:
        return [sum(nums) - sum(set(nums)), sum(range(1, len(nums) + 1)) - sum(set(nums))]


nums = [1, 5, 3, 2, 2, 7, 6, 4, 8, 9]
s = Solution()
print(s.findErrorNums1(nums))
